src/data/nat/totient.lean

```diff
@@ -335,4 +338,142 @@ begin
   rw [sub_mul, one_mul, mul_comm, mul_inv_cancel hp', cast_pred hp],
 end
 
+lemma gcd_self_pow (p n m : ℕ) : (p ^ n).gcd (p ^ m) = p ^ (min n m) :=
+begin
+  wlog h : n ≤ m,
+  rw [min_eq_left h],
+  apply gcd_eq_left,
+  exact pow_dvd_pow p h,
+end
+
+lemma totient_pow_mul_self {p : ℕ} (n m : ℕ) (h : nat.prime p) :
+   φ ((p ^ n).gcd (p ^ m)) * φ (p ^ n * p ^ m) = φ (p ^ n) * φ (p ^ m) * (p ^ n).gcd (p ^ m) :=
+begin
+  wlog hnm : n ≤ m using [n m],
+  rcases n.eq_zero_or_pos with rfl | hn,
+  { simp only [nat.gcd_one_left, mul_one, one_mul, pow_zero] },
+  rcases m.eq_zero_or_pos with rfl | hm,
+  { simp only [mul_one, one_mul, nat.gcd_one_right, totient_one, pow_zero] },
+  have h20 : 0 < n + m, by linarith,
+  rw [totient_prime_pow h hn, totient_prime_pow h hm, ←pow_add, totient_prime_pow h h20,
+      gcd_self_pow, min_eq_left hnm, totient_prime_pow h hn],
+  have : p ^ (n + m - 1) * (p - 1) = (p ^ (m - 1) * (p - 1)) * p^n ,
+  by { have : p^(n+m-1) = p^n * p^(m-1), by {rw nat.add_sub_assoc, apply pow_add, linarith, },
+  rw [this, mul_rotate], },
+  rw [this, ←mul_assoc],
+end
+
+lemma p_val_nat_div_coprime (a p : ℕ) (ha : a ≠ 0 ) (hp : nat.prime p) :
+  coprime p (a/ p^(padic_val_nat p a)) :=
+begin
+  haveI : fact p.prime, exact {out := hp},
+  rw padic_val_nat_eq_factorization p a,
+  apply nat.coprime_of_div_pow_factorization hp (ha),
+end
+
+/-- An induction principle. -/
+def strong_sub_induction {P : ℕ → ℕ → Sort u}
+  (H : ∀ n m, ((∀ x y, x < n → y < m → P x y) → P n m)) : Π (n m : ℕ), P n m :=
+begin
+  intros n m,
+  apply H _ _,
+  induction n,
+  { intros x y hx hy, simp only [not_lt_zero'] at hx, cases hx },
+  { intros x y H1 H2,
+  apply or.by_cases (decidable.lt_or_eq_of_le (le_of_lt_succ H1)),
+  refine λ hn, (n_ih _ _ hn H2),
+  intro hn,
+  rw hn,
+  refine H n_n y (λ a b hab hab2, n_ih a b hab (lt_trans hab2 H2)) },
+end
+
+lemma totient_mul_gen : ∀ a b: ℕ, φ (a.gcd b) * φ (a * b) = φ a * φ b * (a.gcd b) :=
+begin
+  apply strong_sub_induction,
+  intros n m hxy,
+  by_cases hco : coprime n m,
+  rw coprime_iff_gcd_eq_one.1 hco,
+  simp only [totient_one, one_mul, mul_one, totient_mul hco],
+  by_cases g1 : n ≠ 0,
```

Comment:
You should open some brackets here, otherwise it's not obvious what the scope of this assumption is. Same on the line below.

Also, you might want to use `rcases eq_or_ne n 0 with rfl | g1` as that rewrites `n = 0` automatically for you.